### PR TITLE
Issue #4247 SameSite Session Cookie

### PR DESCRIPTION
**Pull request by gregw**
#4247
Allows sameSite cookie settings to be configured in SessionCookieConfig comments

### CODE REVIEW

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpCookie.java [hunk 1]
```diff
@@ -465,6 +465,41 @@ public static String getCommentWithoutAttributes(String comment)
         return strippedComment.length() == 0 ? null : strippedComment;
     }
 
+    public static String getCommentWithAttributes(String comment, boolean httpOnly, SameSite sameSite)
```
Comment (janbartel) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpCookie.java]: Is there a test for this new method?

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java [hunk 2]
```diff
@@ -530,6 +530,16 @@ public boolean getHttpOnly()
         return _httpOnly;
     }
 
+    /**
+     * @return The sameSite setting for session cookies or null for no setting
+     * @see HttpCookie#getSameSite()
+     */
+    @ManagedAttribute("SameSite setting for session cookies")
+    public HttpCookie.SameSite getSameSite()
```
Comment (janbartel) [inline jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: I'd have been fine with a protected field, but doing it in the comment is fine.

**Comment by joakime** [comment]
This is still in Draft mode.

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java [hunk 3]
```diff
@@ -650,30 +660,18 @@ public HttpCookie getSessionCookie(HttpSession session, String contextPath, bool
...
+            cookie = new HttpCookie(
+                _cookieConfig.getName(),
+                id,
+                _cookieConfig.getDomain(),
+                sessionPath,
+                _cookieConfig.getMaxAge(),
+                _cookieConfig.isHttpOnly(),
+                _cookieConfig.isSecure() || (isSecureRequestOnly() && requestIsSecure),
+                HttpCookie.getCommentWithoutAttributes(_cookieConfig.getComment()),
+                1,
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: Is this `1` for version 1 cookies?
That's deprecated.

Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: This was a question. :nerd_face: 
Didn't mean for you to blindly change it to 0.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpCookie.java (hunk 1 above)
Comment (joakime) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpCookie.java]: Yes, this PR needs tests.

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java (hunk 3 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: The old code used "1" if there was a Session Comment, and didn't use "1" (not sure what default value is) if there wasn't a Session Comment.

https://github.com/eclipse/jetty.project/blob/e09444eeb52c3454c4b78cd27ec47187fc44700e/jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java#L653-L676

**Comment by joakime** [comment]
Can we have a jetty-home module property to set this SameSite value for session handling somehow?

**Comment by gregw** [comment]
@joakime not really because it is a per-webapp configuration not per-server

**Comment by joakime** [comment]
Then how about a configuration that can be put into the `WEB-INF/web.xml`?

Such as ...

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<web-app>
  <context-param>
    <param-name>org.eclipse.jetty.session.Cookie.SameSite</param-name>
    <param-value>LAX</param-value>
  </context-param>
</web-app>
```

That way users can set the option in their WAR files, and when running on Jetty the option kicks in for that webapp only?

**Comment by gregw** [comment]
@joakime The `SessionCookieConfig` mechanism is the standard way to configure session cookies, so I don't want to add another way to do it.   Eventually there will be API for sameSite added to that interface and then we can just change the comment code to the new API in the same area, without needing to look elsewhere.  For embedded usage, we do have new API already

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java (hunk 3 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: I think the whole usage of versioned cookies is deprecated, so I don't think we should every use version 1, even if we have comments.

Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: Versioned cookies are used.
Old installations still use it.
And the Servlet TCK mandates it.

Comment (olamy) [reply jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: correct TCK test use version 1. I will run a tck test with the branch to see what happen.

Comment (olamy) [reply jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: fyi doesn't break any cookie tck tests (https://jenkins.webtide.net/job/nightlies/job/servlettck-run-jetty-9.4.x/323/) please note tck tests are run using 
```
jetty.httpConfig.requestCookieCompliance=RFC2965
jetty.httpConfig.responseCookieCompliance=RFC2965
```

**Comment by gregw** [comment]
@joakime I have changed the comment lookup so that precedence goes `STRICT` > `LAX` > `None`
However, with regards to forcing Secure, I do not think we should be enforcing browser best practises, which may change.  Users are free to configure `Secure` + `None`.

**Comment by joakime** [comment]
The explained rationale is ...

The reason `SameSite` exists is as a protection against CSRF and https://tools.ietf.org/html/rfc7258
If you have `SameSite=None`, and **no** `Secure`, then you are vulnerable again.

**Pull request merged**